### PR TITLE
CTX-5524: Added tag management and inference.py hotfix

### PR DESCRIPTION
**Pull request by VukManojlovic**

### CODE REVIEW

File: coretex/entities/tag.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+            "name": tagName,
+            "type": int(self.entityTagType),
+            "project_id": self.projectId,
+        }
+
+        response = networkManager.get("tag", parameters)
+        if response.hasFailed():
+            raise NetworkRequestError(response, "Failed to check existing tags")
+
+        tags = response.getJson(dict)["data"]
```
Comment (dule1322) [inline coretex/entities/tag.py:56]: Check if `data` field exists and is of correct type when accessing it.

File: coretex/entities/tag.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        if len(tags) == 0:
+            return None
+
+        tagId: int = tags[0]["id"]
```
Comment (dule1322) [inline coretex/entities/tag.py:60]: Check if `id` field exists and is of correct type when accessing it.

File: coretex/entities/tag.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    @property
+    @abstractmethod
+    def entityTagType(self) -> int:
+        pass
+
+    def _getTagId(self, tagName: str) -> Optional[int]:
+        parameters = {
+            "name": tagName,
+            "type": int(self.entityTagType),
+            "project_id": self.projectId,
```
Comment (dule1322) [inline coretex/entities/tag.py:49]: Remove `,` from the last member of dict

Comment (dule1322) [inline coretex/entities/tag.py:49]: Why is `project_id` needed to fetch tags?

**Review by dule1322 — CHANGES_REQUESTED** [review]

File: coretex/entities/tag.py (hunk 3 above)
Comment (VukManojlovic) [reply coretex/entities/tag.py:49]: This way only tags of the relevant project are returned. We then check if the inputted tag already exists for that project.

File: coretex/entities/tag.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+    @property
+    @abstractmethod
+    def entityTagType(self) -> int:
+        pass
+
+    def _getTagId(self, tagName: str) -> Optional[int]:
+        parameters = {
+            "name": tagName,
+            "type": int(self.entityTagType),
```
Comment (bogdant36) [inline coretex/entities/tag.py:48]: isn't entityTagType already of type int? why is casting to int necessary here?

File: coretex/entities/tag.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            "tags": tags
+        }
+
+        response = networkManager.post("tag/entity", parameters)
+        if response.hasFailed():
+            raise NetworkRequestError(response, "Failed to create tag")
+
+    def removeTag(self, tag: str) -> None:
+        """
+            Remove tag with provided name from the model
```
Comment (bogdant36) [inline coretex/entities/tag.py:126]: does this function removes tag from model entity only?

File: coretex/utils/inference.py
Comment (bogdant36) [inline coretex/utils/inference.py:1]: Why are there changes to inference.py file that are not related to tag management like described in pr title and description. This makes changes hard to track. As far as I've seen CTX-5524 is related to Endpoints - Zero-knowledge ML, I don't see any tag related comments that should be done in scope of this task.

File: coretex/entities/tag.py (hunk 4 above)
Comment (VukManojlovic) [reply coretex/entities/tag.py:48]: Backend receives a EntityTagType class for the get request (which doesn't happen for post requests) so I cast it to int even though it behaves like an integer object in python code.

File: coretex/utils/inference.py
Comment (VukManojlovic) [reply coretex/utils/inference.py:1]: CTX-5524 is the first task that requests tags to be added programmatically, so the coretexpylib code for that was also included as part of this task. I will add a comment that will clarify this.

inference.py was not initially included in the PR, and the change done to it is a minor hotfix for verified inference, so the title never changed. I see now that the title may be confusing, so I will make it more accurate for the PR.

File: coretex/entities/tag.py (hunk 4 above)
Comment (dule1322) [reply coretex/entities/tag.py:48]: Why not just change the return type of `entityTagType` to `EntityTagType` and then instead of casting it to `int` here you can do it like `self.entityTagType.value`.

Comment (VukManojlovic) [reply coretex/entities/tag.py:48]: The exact value return value of `EntityTagType.dataset` is `Literal[EntityTagType.dataset]` and similarly `.model` for model. How about we instead keep `int` as the return type of `entityTagType` and return EntityTagType.dataset.value / .model.value.
This way we keep a standardized return value for `entityTagType`. All uses of this property need the enum integer value anyway.

Comment (dule1322) [reply coretex/entities/tag.py:48]: But that would allow returning any integer, it's better to use enums to limit the number of available values which can be returned to prevent potential errors during coding. If you put `int` as the return type you can return any kind of integer for that, whilst if the return type of the property is `EntityTagType` mypy will complain if you do not return a valid value from that enum.

**Review by dule1322 — APPROVED** [review]
